File: robocode-tankroyale-server/src/main/java/dev/robocode/tankroyale/server/mappers/BulletToBulletStateMapper.java
package dev.robocode.tankroyale.server.mappers;

import dev.robocode.tankroyale.server.model.Bullet;
import dev.robocode.tankroyale.server.model.Point;
import dev.robocode.tankroyale.server.util.MathUtil;
import dev.robocode.tankroyale.schema.BulletState;

@SuppressWarnings("WeakerAccess")
public final class BulletToBulletStateMapper {

	private BulletToBulletStateMapper() {}

	public static BulletState map(Bullet bullet) {
		BulletState bulletState = new BulletState();

		double direction = bullet.getDirection();
		double speed = bullet.getSpeed();

		bulletState.setOwnerId(bullet.getBotId());
		bulletState.setBulletId(bullet.getBulletId());
		bulletState.setDirection(MathUtil.normalAbsoluteDegrees(direction));
		bulletState.setPower(bullet.getPower());
		bulletState.setSpeed(speed);
		Point pos = bullet.calcPosition();
		bulletState.setX(pos.x);
		bulletState.setY(pos.y);

		return bulletState;
	}
}

File: robocode-tankroyale-server-kotlin/src/main/kotlin/dev/robocode/tankroyale/server/mapper/BulletToBulletStateMapper.kt
package dev.robocode.tankroyale.server.mapper

import dev.robocode.tankroyale.schema.BulletState
import dev.robocode.tankroyale.server.math.normalAbsoluteDegrees
import dev.robocode.tankroyale.server.model.IBullet

object BulletToBulletStateMapper {
    fun map(bullet: IBullet): BulletState {
        return BulletState().apply {
            ownerId = bullet.botId.id
            bulletId = bullet.bulletId.value
            direction = normalAbsoluteDegrees(bullet.direction)
            power = bullet.power
            speed = bullet.speed()
            x = bullet.position().x
            y = bullet.position().y
            color = bullet.color?.value
        }
    }
}